static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/python3
+
+import os
+import re
+import tempfile
+import yaml
+import contextlib
+from pathlib import Path
+
+from lib import util, results
+
+
+# Obtained from
+# https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#task
+ansible_reserved_keywords = [
+    'action', 'any_errors_fatal', 'args', 'async', 'become', 'become_exe', 'become_flags',
+    'become_method', 'become_user', 'changed_when', 'check_mode', 'collections', 'connection',
+    'debugger', 'delay', 'delegate_facts', 'delegate_to', 'diff', 'environment', 'failed_when',
+    'ignore_errors', 'ignore_unreachable', 'local_action', 'loop', 'loop_control',
+    'module_defaults', 'name', 'no_log', 'notify', 'poll', 'port', 'register', 'remote_user',
+    'retries', 'run_once', 'tags', 'throttle', 'timeout', 'until', 'vars', 'when', 'with_items'
+]
+
+
+@contextlib.contextmanager
+def select_all_rules(datastream):
+    """
+    Given an XML file path as 'datastream' enable all the rules in the datastream.
+
+    Use as a context manager, returning a filesystem path to the
+    modified datastream.
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules selected as {new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def traverse_blocks(task, allowed_modules):
+    block_set = set(['block'])
+    task_keywords = set(task.keys())
+
+    if len(task_keywords.intersection(block_set)) > 0:
+        for block_task in task['block']:
+            traverse_blocks(block_task, allowed_modules)
+    else:
+        updated_keywords = set()
+        for keyword in task_keywords:
+            keyword = keyword.replace('ansible.builtin.', '')
+            updated_keywords.add(keyword)
+        task_module = list(updated_keywords.intersection(allowed_modules))
+        if len(task_module) == 1:
+            return str(task_module[0])
+        else:
+            module = ''.join(
+                [x for x in list(updated_keywords) if x not in ansible_reserved_keywords]
+            )
+            results.report('fail', module, 'module is not allowed')
+            util.log(f"This occurred in the following task:\n{task}\n")
+            return None
```

Comment:
This part is confusing.

Comparing it to https://src.fedoraproject.org/tests/scap-security-guide/blob/main/f/Sanity/ansible-allowed-modules/check_allowed_modules.py#_6 there is nothing about reserved keywords. Why is it needed there? And why is it used to filter our keywords right before reporting `fail`? What if `task_module` has multiple keywords, one of them is allowed module and others are reserved keywords? In that case this false positive.

Reply:
It is there to identify name of the module which is not allowed. When you won't find any allowed module you don't know which item from the python dictionary represents a valid module name and what is just Ansible keyword (used in tasks, e.g. `when`, `tags`, etc.). Therefore, we have a list of all allowed keywords in tasks and we filter them out which will leave us with one item and that should be the name of the module.

Reply:
I will add a comment with explanation.